testapi.pm

```diff
@@ -1202,14 +1202,22 @@ alternatively matches a regular expression. Use it as
 
   validate_script_output "cat /etc/hosts", sub { m/127.*localhost/ };
   validate_script_output "cat /etc/hosts", qr/127.*localhost/;
+  validate_script_output "cat /etc/hosts", qr/127.*localhost/, title => 'localhost check', fail_message => 'No localhost line in /etc/hosts!';
   validate_script_output "cat /etc/hosts", sub { $_ !~ m/987.*somehost/ };
 
 =cut
 
 sub validate_script_output {    # no:style:signatures
     my ($script, $check, @args) = @_;
+    my %args = compat_args(
+        {
+            title              => 'validate_script_output',
+            timeout            => undef,
+            fail_message       => 'output not validating',
+            quiet              => 1, # less noise by default
+        }, ((@args % 2) == 0) ? [] : ['timeout'], @args);
 
-    my $output = script_output($script, @args);
+    my $output = $distri->script_output($script, $args{timeout}, %args);
```

Comment:
Could you explain why this change is needed? i dont think is wrong but i would like to know what is the reasoning behind it?

Reply:
The parameter passing does not work if `compat_args` is called twice because of `$timeout` as fixed first parameter.
Via calling `$distri->script_output` directly I avoid that issue (and have less function calls).

Reply:
I think the problem is, that you pass parameters like `title` and `fail_message` in the positional parameter slot, thus `compat_args` can not judge if these are named or positional and take it as positional parameter. This could result in the name `title` used as `timeout`.

If you call it like you did, you could also use the `testapi::script_output()` call. And I would prefer to use it, because the for instance the `quite` parameter is taken from `_QUITE_SCRIPT_CALLS` which isn't in the distribution!

```suggestion
    my $output = script_output($script, $args{timeout}, %args);
```

Reply:
What I am not sure about @cfconrad 's workaround is that there are calls to `compat_args()` both in `testapi::script_output()` and in `distribution::script_output()` neither of which have a previous check for the number of arguments in `@_`. If `testapi::script_output()` is already using `compat_args()` to modify positional arguments into named arguments, then doing the same again in `distribution::script_output()` doesn't make much sense to me. I think we've been lucky that this has been working as is.

Reply:
That is actually the idea of `compat_args()` You can call a function like:
```
func($a, $b, $b);
```perl
and 
```
func('foo' => $a, 'bla' => $b, 'blub' => $c);
```
But also as:
```perl
func($a, $b);
func($a);
func(bla=>$b);
```

And both result in the same behavior as long as you specify compat_args like
```perl
my %args = compat_args(
        {
            foo => 'something',
            bla => 'is',
            blub => 'this',
        }, ['foo', 'bla', 'blub'], @args);
```
see: https://github.com/os-autoinst/os-autoinst/blob/master/t/03-testapi.t#L661

And `compat_args` get problems when we pass not specified arguments in the positional one. For the example above it would:
```perl
func(k => $bob);
```
In that case, `compat_args` do not recognize the key `k` as named parameter, thus it take it as positional for `foo` and the value `$bob` as positional for `bla`... I think we run in this trap here as well, cause we extend the arguments from `script_output()` but do not specify them in the `compat_args` call in `validate_script_output`.

Reply:
There would be the case that we can change `compat_args` in that regards, if there are only one positional parameter, we could make some extra judgement to get that solved. But I do not see it with 2 or more positional parameters.

Reply:
> And `compat_args` get problems when we pass not specified arguments in the positional one. For the example above it would:
> 
> ```perl
> func(k => $bob);
> ```
> 
> In that case, `compat_args` do not recognize the key `k` as named parameter, thus it take it as positional for `foo` and the value `$bob` as positional for `bla`... I think we run in this trap here as well, cause we extend the arguments from `script_output()` but do not specify them in the `compat_args` call in `validate_script_output`.

That was the bit I was missing, and it explains why we have not had issues with the 2 inner calls to `compat_args()` in `testapi::script_output()` and `distribution_script_output()` as both `compat_args()` calls deal with the same 4 named args (timeout, proceed_on_failure, quiet and type_command), and why it's causing @bs-suse some troubles when any of his new named args reach the inner calls.

@bs-suse I maintain that the new named args are not required by either `testapi::script_output()` or by `distribution::script_output()`, but now I suggest instead of having the `compat_args()` call be placed after the call to `script_output()`, you can drop from the hash those named args you introduced that are not used by `script_output()`, so my suggestion would be:

```
sub validate_script_output {    # no:style:signatures
    my ($script, $check, @args) = @_;
    my %args = compat_args(
        {
            title => 'validate_script_output',
            timeout => undef,
            fail_message => 'output not validating',
            quiet => 1,    # less noise by default
        }, ((@args % 2) == 0) ? [] : ['timeout'], @args);
my $title = $args{title};
my $fail_message = $args{fail_message};
delete $args{title};
delete $args{fail_message};
my $output = script_output($script, %args);
```
And in the rest of the method, you refer to `$title` and `$fail_message` instead of `$args{title}` and `$args{fail_message}`.

Reply:
I think if we will add this "one fixed parameter" case in `compat_args` you do not need to change the call at all. Also this `(@args % 2) == 0` is then not needed anymore. See:
https://github.com/os-autoinst/os-autoinst/pull/2169